core/sched/rt-rm.c

```diff
@@ -0,0 +1,233 @@
+
+#ifdef SCHED_RM_REPORT
+    /* for analysis */
+    uint32_t foreground_ticks;
+    uint32_t preempted;
+#endif
+
+    struct list_head head;
+};
+
+struct sched_data_rm {
+    struct scheduler *s;
+
+    uint64_t tick_interval_us;
+    uint64_t tick_count;
+#ifdef SCHED_RM_REPORT
+    uint64_t tick_1kcount;
+#endif
+
+    struct entry_data_rm *current;
+    struct list_head runqueue;
+};
+
+/* Scheduler related data initialization */
+void sched_rm_init(struct scheduler *s)
+{
+    struct sched_data_rm *sd = (struct sched_data_rm *) malloc(sizeof(struct sched_data_rm));
+
+    /* Check scheduler config */
+    /* Allocate memory for system-wide data */
+    /* Initialize data */
+    sd->current = NULL;
+    sd->tick_count = 0;
+#ifdef SCHED_RM_REPORT
+    sd->tick_1kcount = 1000;
+#endif
+    sd->tick_interval_us = schedconf_rm_tick_interval_ms[s->pcpuid];
+    sd->s = s;
+
+    LIST_INITHEAD(&sd->runqueue);
+
+    s->sd = sd;
+}
+
+int sched_rm_vcpu_register(struct scheduler *s, struct sched_entry *e)
+{
+    // struct sched_data_rm *sd = (struct sched_data_rm *) (s->sd);
+    struct entry_data_rm *ed = (struct entry_data_rm *) malloc(sizeof(struct entry_data_rm));
+
+    ed->period = schedconf_rm_period_budget[e->vcpuid][0];
+    ed->budget = schedconf_rm_period_budget[e->vcpuid][1];
+    ed->e = e;
+
+#ifdef SCHED_RM_REPORT
+    ed->foreground_ticks = 0;
+    ed->preempted = 0;
+#endif
+
+    LIST_INITHEAD(&ed->head);
+
+    e->ed = ed;
+
+    return 0;
+}
+
+int sched_rm_vcpu_unregister(struct scheduler *s, struct sched_entry *e)
+{
+    /* Check if vcpu is registered */
+    /* Check if vcpu is detached. If not, request detachment.*/
+    /* If we have requested detachment of vcpu,
+     *   let's wait until it is detached by main scheduling routine */
+
+    return 0;
+}
+
+int sched_rm_vcpu_attach(struct scheduler *s, struct sched_entry *e)
+{
+    struct sched_data_rm *sd = (struct sched_data_rm *) (s->sd);
+    struct entry_data_rm *ed = (struct entry_data_rm *) (e->ed);
+
+    ed->period_cntdn = ed->period;
+    ed->budget_cntdn = ed->budget;
+
+    LIST_ADDTAIL(&ed->head, &sd->runqueue);
+
+    return 0;
+}
+
+int sched_rm_vcpu_detach(struct scheduler *s, struct sched_entry *e)
+{
+    // struct sched_data_rm *sd = (struct sched_data_rm *) (s->sd);
+    struct entry_data_rm *ed = (struct entry_data_rm *) (e->ed);
+
+    /* Check if vcpu is attached */
+    /* Remove it from runqueue by setting will_detached flag*/
+    /* Set entry_data_rm's fields */
+
+    LIST_DELINIT(&ed->head);
+
+    return 0;
+}
+
+/**
+ * Main scheduler routine in RT - Rate Monotonic policy implementation
+ */
+int sched_rm_do_schedule(struct scheduler *s, uint64_t *expiration)
+{
+    struct sched_data_rm *sd = (struct sched_data_rm *) (s->sd);
+
+    struct entry_data_rm *current_ed = NULL;
+    struct entry_data_rm *next_ed = NULL;
+
+    // bool is_switching_needed = false;
+    int next_vcpuid = VCPUID_INVALID;
+
+    sd->tick_count += 1;
```

Comment:
For unity, you might want to changed tick_count to tick_cnt.
you have an other option to changed xxx_cntdn to xxx_countdown